### DISCUSSION TITLE
Is there a way to upload a file and download it in a workflow?

### DISCUSSION THREAD

**Discussion by OmarJalil** [codemagic-ci-cd/codemagic-docs#2625] [Q&A] [answered]
### If you are looking for help, please confirm the following...

- [X] I confirm I have searched the [Docs](https://docs.codemagic.io), [Codemagic Sample Projects](https://github.com/codemagic-ci-cd/codemagic-sample-projects), and GitHub [Discussions](https://github.com/orgs/codemagic-ci-cd/discussions).
- [X] I confirm the software versions on my local computer match the ones in CI environment (e.g. Xcode version, Gradle version). See [here](https://docs.codemagic.io/troubleshooting/common-android-issues/#builds-work-locally-but-fail-on-codemagic)
- [X] I confirm I am able to do a fresh clone of the repository and build the project on my local computer.

### Which mobile framework are you using?

iOS (Swift, Objective-C)

### Steps to reproduce

I want to be able to download a let say .env file. 

### Expected results

I want to be able to upload it directly to CodeMagic and download it later with a worflow

### Actual results

I cannot do that

**Comment by dtrdic** [comment, marked as answer]
Hi @OmarJalil, what is your use case here? Are you looking to import this file from a secure S3 bucket and make it downloadable on the UI, or do you want to use values from that file within your workflow?

In any case you would need to upload this.env file somewhere in order to import it, then write the values to the `CM_ENV` environment variable.
```
scripts:
  - name: Get assets from AWS S3 bucket
    script: |
      aws s3 cp s3://$S3_BUCKET_NAME.zip assets.zip
      unzip assets.zip -d $ASSETS_FOLDER
      ls $ASSETS_FOLDER
  - name: Set iOS environment variables from settings.env
    script: |
      source $HOME/clone/$ASSETS_FOLDER/settings.env

      echo "APP_STORE_CONNECT_KEY_IDENTIFIER=$APP_STORE_CONNECT_KEY_IDENTIFIER" >> $CM_ENV
      echo "APP_STORE_CONNECT_ISSUER_ID=$APP_STORE_CONNECT_ISSUER_ID" >> $CM_ENV

      echo "APP_STORE_CONNECT_PRIVATE_KEY<<DELIMITER" >> $CM_ENV
      echo "$APP_STORE_CONNECT_PRIVATE_KEY" >> $CM_ENV
      echo "DELIMITER" >> $CM_ENV

      echo "CERTIFICATE_PRIVATE_KEY<<DELIMITER" >> $CM_ENV
      echo "$CERTIFICATE_PRIVATE_KEY" >> $CM_ENV
      echo "DELIMITER" >> $CM_ENV
```
Or, if you just want this file to be downloadable from the UI, you can do it like this:

```
        scripts:   
          - name: Get assets from AWS S3 bucket
            script: |
              FILE_PATH="sample.txt"
              S3_BUCKET_NAME: test-bucket
              DESTINATION_FOLDER="$CM_BUILD_DIR/ASSETS_FOLDER"

              # Create the destination folder if it doesn't exist
              mkdir -p $DESTINATION_FOLDER

              # Copy the file from S3 bucket to the local destination
              aws s3 cp s3://$S3_BUCKET_NAME/$FILE_PATH $DESTINATION_FOLDER

              # List the files in the destination folder to verify
              ls $DESTINATION_FOLDER
              
         artifacts:
          - $CM_BUILD_DIR/ASSETS_FOLDER/*.txt //add this path to be able to download the file from the UI
```

**Reply by OmarJalil** [reply]
That could also help! thank you for that, but we currently use Bitrise and I was looking for another option, bitrise allows you to upload files and use them with a dedicated step to download the file. This allows the dev team to not depend on replacing the file on an S3 bucket, but just to drag and drop a file inside the platform, reducing friction and implementation times. I think this can be a good addition to CodeMagic 

https://devcenter.bitrise.io/en/builds/managing-build-files/using-files-in-your-builds.html
https://devcenter.bitrise.io/en/builds/managing-build-files/uploading-files-for-your-builds.html

**Reply by dtrdic** [reply]
Thank you for the feedback, @OmarJalil. I'll classify this use case as a feature request, and the team will evaluate it for potential improvement in future iterations. I'll close this thread as resolved since there's a workaround for it. 
Have a nice day!
